File: app/src/server/progress.ts
import type { ProgressEvent } from '@shared/types';

export function eventToSse(event: ProgressEvent): string {
  return `event: ${event.kind}\ndata: ${JSON.stringify(event.data)}\n\n`;
}

interface Subscription {
  queue: ProgressEvent[];
  resolve: (() => void) | undefined;
}

export class EventBus {
  private subscribers = new Set<Subscription>();

  async publish(event: ProgressEvent): Promise<void> {
    for (const sub of this.subscribers) {
      sub.queue.push(event);
      sub.resolve?.();
    }
  }

  async *subscribe(): AsyncIterable<ProgressEvent> {
    const sub: Subscription = { queue: [], resolve: undefined };
    this.subscribers.add(sub);
    try {
      while (true) {
        const next = sub.queue.shift();
        if (next !== undefined) {
          yield next;
          continue;
        }
        await new Promise<void>((r) => {
          sub.resolve = r;
        });
        sub.resolve = undefined;
      }
    } finally {
      this.subscribers.delete(sub);
    }
  }
}

File: src/notion_extractor/progress.py
import asyncio
import enum
import json
from collections.abc import AsyncIterator
from dataclasses import dataclass, field
from typing import Any


class EventKind(enum.StrEnum):
    DISCOVERY_STARTED = "discovery_started"
    DISCOVERY_PROGRESS = "discovery_progress"
    DISCOVERY_DONE = "discovery_done"
    NODE_STARTED = "node_started"
    NODE_DONE = "node_done"
    NODE_FAILED = "node_failed"
    ATTACHMENT_DOWNLOADED = "attachment_downloaded"
    EXTRACTION_DONE = "extraction_done"
    ERROR = "error"


@dataclass
class Event:
    kind: EventKind
    data: dict[str, Any] = field(default_factory=dict)

    def to_sse(self) -> str:
        payload = json.dumps(self.data)
        return f"event: {self.kind.value}\ndata: {payload}\n\n"


class EventBus:
    def __init__(self) -> None:
        self._subscribers: list[asyncio.Queue[Event]] = []

    async def publish(self, event: Event) -> None:
        for q in list(self._subscribers):
            await q.put(event)

    async def subscribe(self) -> AsyncIterator[Event]:
        queue: asyncio.Queue[Event] = asyncio.Queue()
        self._subscribers.append(queue)
        try:
            while True:
                yield await queue.get()
        finally:
            self._subscribers.remove(queue)
